hs_core/models.py

```diff
@@ -1469,6 +1469,181 @@ def supports_folders(self):
+        if self.metadata is None or not self.metadata.has_all_required_elements():
+            return False
+        for f in self.logical_files:
+            if not f.metadata.has_all_required_elements():
+                return False
+        return True
+
+    @property
+    def can_be_public_or_discoverable(self):
+        """
+        Return True if the resource can be set to public or discoverable
+
+        This is True if
+
+        1. The resource has all metadata elements marked as required.
+        2. The resource has all files that are considered required.
+
+        and False otherwise
+        """
+        has_files = self.has_required_content_files
+        has_metadata = self.has_required_metadata
+        return has_files and has_metadata
+
+    def set_discoverable(self, value, user=None):
+        """
+        Set the discoverable flag for a resource.
+
+        :param value: True or False
+        :param user: user requesting the change, or None for changes that are not user requests.
+        :raises ValidationError: if the current configuration cannot be set to desired state
+
+        This sets the discoverable flag (self.raccess.discoverable) for a resource based
+        upon application logic. It is part of AbstractResource because its result depends
+        upon resource state, and not just access control.
+
+        * This flag can only be set to True if the resource passes basic validations
+          `has_required_metata` and `has_required_content_files`
+        * setting `discoverable` to `False` also sets `public` to `False`
+        * setting `discoverable` to `True` does not change `public`
+
+        Thus, the setting public=True, discoverable=False is disallowed.
+
+        If `user` is None, access control is not checked.  This happens when a resource has been
+        invalidated outside of the control of a specific user. In this case, user can be None
+        """
+        # access control is separate from validation logic
+        if user is not None and not user.uaccess.can_change_resource_flags(self):
+            raise ValidationError("You don't have permission to change resource sharing status")
+
+        # check that there is sufficient resource content
+        has_metadata = self.has_required_metadata
+        has_files = self.has_required_content_files
+        if value and not (has_metadata or has_files):
+
+            if not has_metadata and not has_files:
+                msg = "Resource does not have sufficient metadata and content files to be " + \
+                    "discoverable"
+                raise ValidationError(msg)
+            elif not has_metadata:
+                msg = "Resource does not have sufficient metadata to be discoverable"
+                raise ValidationError(msg)
+            elif not has_files:
+                msg = "Resource does not have sufficient content files to be discoverable"
+                raise ValidationError(msg)
```

Comment:
@alvacouch I believe your replacement is not sufficient since it requires ```has_metadata``` and ```has_files``` to be all ```False``` which is not what we want, right? We need either ```has_metadata``` is false or ```has_files``` is false which will suffice using the logic I outlined above.